__tests__/stateProvider.test.ts

```diff
@@ -0,0 +1,71 @@
+import * as core from "@actions/core";
+
+import { Events, RefKey, State } from "../src/constants";
+import {
+    IStateProvider,
+    NullStateProvider,
+    StateProvider
+} from "../src/stateProvider";
+
+jest.mock("@actions/core");
+
+beforeAll(() => {
+    jest.spyOn(core, "getInput").mockImplementation((name, options) => {
+        return jest.requireActual("@actions/core").getInput(name, options);
+    });
+
+    jest.spyOn(core, "setOutput").mockImplementation((key, value) => {
+        return jest.requireActual("@actions/core").setOutput(key, value);
+    });
+});
+
+afterEach(() => {
+    delete process.env[Events.Key];
+    delete process.env[RefKey];
+});
+
+test("StateProvider saves states", async () => {
+    const getStateMock = jest
+        .spyOn(core, "getState")
+        .mockImplementation(name =>
+            jest.requireActual("@actions/core").getState(name)
+        );
+    const saveStateMock = jest
+        .spyOn(core, "saveState")
+        .mockImplementation((key, value) => {
+            return jest.requireActual("@actions/core").saveState(key, value);
+        });
+
+    const cacheMatchedKey = "node-cache";
+
+    const stateProvider: IStateProvider = new StateProvider();
+    stateProvider.setState("stateKey", "stateValue");
+    stateProvider.setState(State.CacheMatchedKey, cacheMatchedKey);
+    stateProvider.getState("stateKey");
```

Comment:
Should validate that the value fetched is same as value set in L42

Reply:
I tried that, but unfortunately the values stored in the `state` are not accessible in the same step as we discovered last time, hence can't validate it. :/

Reply:
Ah yes, probably we should mock `core.setState` and `core.getState` to achieve this. Currently you are still loading core and using that. May be instead just save in a memory hash and then serve it back with getState. Should be fine to do so as we are not testing `core` functionality here.

But for now it is okay. You can take this if you want in separate PR.

Reply:
Yes, will take this in a separate PR.

Reply:
Raised a PR for this - https://github.com/actions/cache/pull/1029. 
You were looking for something like this right? 🤔